pkg/controller/controllerutil/controllerutil.go

```diff
@@ -188,20 +165,17 @@ func HasOwnerReference(ownerRefs []metav1.OwnerReference, obj client.Object, sch
 	if err != nil {
 		return false, err
 	}
-	idx := indexOwnerRef(ownerRefs, metav1.OwnerReference{
-		APIVersion: gvk.GroupVersion().String(),
-		Name:       obj.GetName(),
-		Kind:       gvk.Kind,
-	})
+	idx := indexOwnerRef(ownerRefs, *NewOwnerRef(obj, gvk))
 	return idx != -1, nil
 }
 
 // RemoveControllerReference removes an owner reference where the controller
 // equals true
 func RemoveControllerReference(owner, object metav1.Object, scheme *runtime.Scheme) error {
-	if ok := HasControllerReference(object); !ok {
-		return fmt.Errorf("%T does not have a owner reference with controller equals true", object)
+	if !metav1.IsControlledBy(object, owner) {
```

Comment:
This error is now misleading because if the object doesn’t have a controller reference the returned error will only indicate that the owner being passed is not the controller reference for the object. 

The error before allowed you to recognize if the object you are passing in had one at all before trying to remove it with the owner passed in.

Reply:
https://github.com/kubernetes-sigs/controller-runtime/pull/2595#discussion_r1404608398 Discussion for why the logic exists was from several issues with how to get around overwriting and errors that exist with controller reference/owner reference.

Reply:
okay. gonna change it soon

Reply:
@troy0820 WDYT about this comment: https://github.com/kubernetes-sigs/controller-runtime/pull/2958#issuecomment-2368877244